File: apps/relay-server/src/routes/calls.py
"""통화 시작/종료 API 엔드포인트.

PRD 8.2:
  POST /relay/calls/start — 전화 발신 + Realtime Session 시작
  POST /relay/calls/{call_id}/end — 통화 종료

통화 시작 시퀀스 (PRD 3.1):
  1. App → Relay Server: POST /relay/calls/start
  2. Relay Server: Twilio 발신 + OpenAI Dual Session 생성
  3. Relay Server → Supabase: call 상태를 CALLING으로 업데이트
  4. Relay Server → App: { relayWsUrl, callSid, sessionIds }
  5. App → Relay Server: WebSocket 연결
"""

import logging

from fastapi import APIRouter, HTTPException

from src.call_manager import call_manager
from src.config import settings
from src.prompt.generator_v3 import generate_session_a_prompt, generate_session_b_prompt
from src.realtime.session_manager import DualSessionManager
from src.tools.definitions import get_tools_for_mode
from src.twilio.outbound import make_call_async
from src.types import (
    ActiveCall,
    CallEndRequest,
    CallStartRequest,
    CallStartResponse,
    CallStatus,
)

router = APIRouter(tags=["calls"])
logger = logging.getLogger(__name__)


@router.post("/calls/start", response_model=CallStartResponse)
async def start_call(req: CallStartRequest):
    """전화 발신을 시작하고 OpenAI Dual Session을 생성한다."""
    if call_manager.get_call(req.call_id):
        raise HTTPException(status_code=409, detail="Call already in progress")

    logger.info(
        "Starting call: id=%s, mode=%s, %s→%s",
        req.call_id,
        req.mode.value,
        req.source_language,
        req.target_language,
    )

    # 1. ActiveCall 생성
    call = ActiveCall(
        call_id=req.call_id,
        mode=req.mode,
        source_language=req.source_language,
        target_language=req.target_language,
        status=CallStatus.CALLING,
        collected_data=req.collected_data or {},
        communication_mode=req.communication_mode,
    )

    # 2. System Prompt 생성
    if req.system_prompt_override:
        prompt_a = req.system_prompt_override
    else:
        prompt_a = generate_session_a_prompt(
            mode=req.mode,
            source_language=req.source_language,
            target_language=req.target_language,
            collected_data=req.collected_data,
        )
    prompt_b = generate_session_b_prompt(
        source_language=req.source_language,
        target_language=req.target_language,
    )

    # Prompt를 ActiveCall에 저장 (AudioRouter에서 사용)
    call.prompt_a = prompt_a
    call.prompt_b = prompt_b

    # 3. OpenAI Dual Session 생성 (vad_mode + communication_mode 전달 — PRD 4.2)
    dual_session = DualSessionManager(
        mode=req.mode,
        source_language=req.source_language,
        target_language=req.target_language,
        vad_mode=req.vad_mode,
        communication_mode=req.communication_mode,
    )

    # Agent Mode: Function Calling 도구 설정
    tools_a = get_tools_for_mode(req.mode.value)

    try:
        await dual_session.connect(prompt_a, prompt_b, tools_a=tools_a)
    except Exception as e:
        logger.error("Failed to create OpenAI sessions: %s", e)
        raise HTTPException(status_code=502, detail="Failed to create AI sessions")

    # 즉시 session 등록 (Twilio 실패 시 cleanup_call로 정리)
    call_manager.register_session(req.call_id, dual_session)

    call.session_a_id = dual_session.session_a.session_id
    call.session_b_id = dual_session.session_b.session_id

    # 4. Twilio 발신 (async)
    try:
        call_sid = await make_call_async(
            phone_number=req.phone_number,
            call_id=req.call_id,
        )
        call.call_sid = call_sid
    except Exception as e:
        logger.error("Failed to make Twilio call: %s", e)
        await call_manager.cleanup_call(req.call_id, reason="twilio_failed")
        raise HTTPException(status_code=502, detail="Failed to initiate phone call")

    # 5. Active call 등록
    call_manager.register_call(req.call_id, call)

    # WebSocket URL 생성
    ws_base = settings.relay_server_url.replace("http", "ws")
    relay_ws_url = f"{ws_base}/relay/calls/{req.call_id}/stream"

    logger.info(
        "Call started: id=%s, sid=%s, ws=%s",
        req.call_id,
        call_sid,
        relay_ws_url,
    )

    return CallStartResponse(
        call_id=req.call_id,
        call_sid=call_sid,
        relay_ws_url=relay_ws_url,
        session_ids={
            "session_a": call.session_a_id,
            "session_b": call.session_b_id,
        },
    )


@router.post("/calls/{call_id}/end")
async def end_call(call_id: str, req: CallEndRequest | None = None):
    """통화를 종료한다."""
    call = call_manager.get_call(call_id)
    if not call:
        raise HTTPException(status_code=404, detail="Call not found")

    reason = req.reason if req else "user_hangup"
    logger.info("Ending call: id=%s, reason=%s", call_id, reason)

    # 중앙 정리 (Twilio 종료 + DB persist 포함)
    await call_manager.cleanup_call(call_id, reason=reason)

    return {"status": "ended", "call_id": call_id, "reason": reason}

File: apps/relay-server/src/routes/calls.ts
import type { FastifyInstance } from 'fastify';
import { config } from '../config.js';
import { initiateOutboundCall } from '../twilio/outbound.js';
import { SessionManager } from '../realtime/session-manager.js';
import { AudioRingBuffer } from '../realtime/ring-buffer.js';
import { SessionRecovery } from '../realtime/session-recovery.js';
import { CostTracker } from '../db/cost-tracker.js';
import { saveCallResult, createCallRecord } from '../db/call-repository.js';
import { ErrorCodes } from '../types.js';
import type { CallStartRequest, CallStartResponse, ActiveCall, CallMode, SessionMode } from '../types.js';
import type { GuardrailEvent } from '../guardrail/checker.js';

// Store active calls (shared with server.ts)
export const activeCalls = new Map<string, {
  call: ActiveCall;
  sessionManager: SessionManager;
  ringBuffer: AudioRingBuffer;
  recovery: SessionRecovery;
  costTracker: CostTracker;
  guardrailEvents: GuardrailEvent[];
}>();

const VALID_MODES: SessionMode[] = ['relay', 'agent'];
const VALID_CALL_MODES: CallMode[] = ['voice-to-voice', 'chat-to-voice', 'voice-to-text'];

export async function callsRoute(app: FastifyInstance) {
  /**
   * POST /relay/calls/start — Start a call with Dual Session
   */
  app.post<{ Body: CallStartRequest }>('/calls/start', async (request, reply) => {
    const { callId, mode, callMode, sourceLanguage, targetLanguage, collectedData } = request.body;

    // Validation
    if (!callId) {
      return reply.status(400).send({
        success: false,
        error: { code: ErrorCodes.MISSING_DATA, message: 'callId is required' },
      });
    }

    if (!VALID_MODES.includes(mode)) {
      return reply.status(400).send({
        success: false,
        error: { code: ErrorCodes.INVALID_MODE, message: `Invalid mode: ${mode}` },
      });
    }

    if (!VALID_CALL_MODES.includes(callMode)) {
      return reply.status(400).send({
        success: false,
        error: { code: ErrorCodes.INVALID_MODE, message: `Invalid callMode: ${callMode}` },
      });
    }

    if (mode === 'agent' && !collectedData) {
      return reply.status(400).send({
        success: false,
        error: { code: ErrorCodes.MISSING_DATA, message: 'collectedData is required for agent mode' },
      });
    }

    // Feature flag check
    if (config.callMode !== 'realtime') {
      return reply.status(400).send({
        success: false,
        error: { code: ErrorCodes.INVALID_MODE, message: 'Realtime mode is not enabled. Set CALL_MODE=realtime' },
      });
    }

    const phone = collectedData?.details?.phone ?? collectedData?.targetPhone;
    if (!phone) {
      return reply.status(400).send({
        success: false,
        error: { code: ErrorCodes.MISSING_DATA, message: 'Target phone number is required' },
      });
    }

    try {
      // 1. Create active call record
      const activeCall: ActiveCall = {
        callId,
        callSid: '',
        mode,
        callMode,
        sourceLanguage,
        targetLanguage,
        collectedData,
        status: 'pending',
        sessionAId: null,
        sessionBId: null,
        startedAt: Date.now(),
        lastActivityAt: Date.now(),
      };

      // 2. Create OpenAI Realtime Dual Sessions
      const sessionManager = new SessionManager(activeCall, {
        onTranscriptUser: () => {},
        onTranscriptUserTranslated: () => {},
        onTranscriptRecipient: () => {},
        onTranscriptRecipientTranslated: () => {},
        onAudioForTwilio: () => {},
        onAudioForApp: () => {},
        onSessionError: (session, error) => {
          app.log.error(`Session ${session} error: ${error}`);
        },
        onRecipientSpeechStarted: () => {},
        onRecipientSpeechEnded: () => {},
        onFirstMessageComplete: () => {},
      });

      const { sessionAId, sessionBId } = await sessionManager.connect();
      activeCall.sessionAId = sessionAId;
      activeCall.sessionBId = sessionBId;

      // 3. Create ring buffer and recovery system
      const ringBuffer = new AudioRingBuffer(30_000);
      const recovery = new SessionRecovery({
        onRecoveryStart: (session) => {
          app.log.warn(`[Recovery] Session ${session} recovery started for call ${callId}`);
        },
        onRecoverySuccess: (session, gapMs) => {
          app.log.info(`[Recovery] Session ${session} recovered, gap: ${gapMs}ms`);
        },
        onRecoveryFailed: (session, error) => {
          app.log.error(`[Recovery] Session ${session} failed: ${error}`);
        },
        onDegradedMode: (session) => {
          app.log.warn(`[Recovery] Session ${session} entering degraded mode`);
        },
      });
      const costTracker = new CostTracker();

      // 4. Initiate Twilio outbound call
      const { callSid } = await initiateOutboundCall({ to: phone, callId });
      activeCall.callSid = callSid;
      activeCall.status = 'calling';

      // 5. Store active call
      activeCalls.set(callId, {
        call: activeCall,
        sessionManager,
        ringBuffer,
        recovery,
        costTracker,
        guardrailEvents: [],
      });

      // 6. Save to DB
      await createCallRecord({
        id: callId,
        user_id: 'system', // TODO: extract from auth
        call_mode: callMode,
        source_language: sourceLanguage,
        target_language: targetLanguage,
        twilio_call_sid: callSid,
        vad_mode: callMode === 'chat-to-voice' ? 'push-to-talk' : 'client',
      });

      // 6. Set call duration limits
      setupCallTimers(callId, activeCall);

      const wsHost = new URL(config.twilioWebhookBaseUrl).host;
      const relayWsUrl = `wss://${wsHost}/relay/calls/${callId}/stream`;

      const response: CallStartResponse = {
        success: true,
        data: {
          callSid,
          sessionA: { id: sessionAId, status: 'connected' },
          sessionB: { id: sessionBId, status: 'connected' },
          relayWsUrl,
          mode,
          callMode,
        },
      };

      return reply.status(200).send(response);
    } catch (err) {
      app.log.error({ err }, 'Failed to start call');

      // Cleanup on failure
      const existing = activeCalls.get(callId);
      if (existing) {
        existing.sessionManager.close();
        activeCalls.delete(callId);
      }

      const isOpenAIError = (err as Error).message?.includes('Session');
      return reply.status(502).send({
        success: false,
        error: {
          code: isOpenAIError ? ErrorCodes.OPENAI_ERROR : ErrorCodes.TWILIO_ERROR,
          message: (err as Error).message,
        },
      });
    }
  });

  /**
   * POST /relay/calls/:id/end — End a call
   */
  app.post<{ Params: { id: string } }>('/calls/:id/end', async (request, reply) => {
    const { id: callId } = request.params;
    const existing = activeCalls.get(callId);

    if (!existing) {
      return reply.status(404).send({
        success: false,
        error: { code: ErrorCodes.CALL_NOT_FOUND, message: 'Active call not found' },
      });
    }

    await endActiveCall(callId, 'completed');

    return reply.status(200).send({ success: true });
  });
}

// ── Call Duration Timers ──

function setupCallTimers(callId: string, call: ActiveCall) {
  // Warning at 8 minutes
  setTimeout(() => {
    const existing = activeCalls.get(callId);
    if (!existing) return;
    // Will be sent via WebSocket stream route
    existing.call.lastActivityAt = Date.now();
  }, config.callWarningAtMs);

  // Auto-end at 10 minutes
  setTimeout(() => {
    const existing = activeCalls.get(callId);
    if (!existing) return;
    console.log(`[CallTimer] Call ${callId} exceeded max duration, auto-ending`);
    endActiveCall(callId, 'completed');
  }, config.maxCallDurationMs);
}

export async function endActiveCall(callId: string, status: 'completed' | 'failed' | 'no_answer') {
  const existing = activeCalls.get(callId);
  if (!existing) return;

  const { call, sessionManager, ringBuffer, recovery, costTracker, guardrailEvents } = existing;

  // Get transcripts before closing
  const transcripts = sessionManager.getTranscripts();
  const durationSeconds = Math.round((Date.now() - call.startedAt) / 1000);

  // Close sessions and cleanup
  sessionManager.close();
  recovery.destroy();
  ringBuffer.clear();

  // End Twilio call if still active
  if (call.callSid) {
    try {
      const { endCall } = await import('../twilio/outbound.js');
      await endCall(call.callSid);
    } catch {
      // Call may already be ended
    }
  }

  // Save to DB with all tracking data
  await saveCallResult(callId, {
    status,
    duration_seconds: durationSeconds,
    transcript_bilingual: transcripts,
    cost_tokens: costTracker.getTokens(),
    guardrail_events: guardrailEvents,
    recovery_events: recovery.getEvents(),
    session_a_id: call.sessionAId,
    session_b_id: call.sessionBId,
  });

  // Remove from active calls
  activeCalls.delete(callId);

  console.log(`[CallManager] Call ${callId} ended with status: ${status}, duration: ${durationSeconds}s`);
}
